File: apps/portal/scripts/refresh_macro_market.py
#!/usr/bin/env python3
"""Refresh GNK ASG macro cross-asset market dataset.

Updates data/macro_market.json for the Bitcoin / Gold / Brent Oil / USD-EUR
comparison panel. Uses public no-key endpoints and keeps the previous dataset
as fallback when one source is temporarily unavailable.
"""

from __future__ import annotations

import json
import math
import sys
import time
import urllib.parse
import urllib.request
from datetime import datetime, timezone
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
DATA = ROOT / "data"
MACRO_PATH = DATA / "macro_market.json"
STATUS_PATH = DATA / "fast_market_status.json"
TIMEOUT = 12
USER_AGENT = "GNK-ASG-MacroMarket/2.0 (+https://gnk-asg.hr/)"

ASSETS = {
    "btc": {"symbol": "BTC-USD", "label": "Bitcoin", "ticker": "BTC", "unit": "USD / BTC", "source": "Yahoo Finance public chart", "invert": False},
    "gold": {"symbol": "GC=F", "label": "Zlato", "ticker": "XAU", "unit": "USD / oz", "source": "Yahoo Finance public chart", "invert": False},
    "oil": {"symbol": "BZ=F", "label": "Brent nafta", "ticker": "BRENT", "unit": "USD / barrel", "source": "Yahoo Finance public chart", "invert": False},
    "usd": {"symbol": "EURUSD=X", "label": "USD / EUR", "ticker": "USD", "unit": "EUR / USD", "source": "Yahoo Finance public chart, inverted EURUSD", "invert": True},
}


def now_iso() -> str:
    return datetime.now(timezone.utc).replace(microsecond=0).isoformat()


def read_json(path: Path, default):
    try:
        if not path.exists() or path.stat().st_size == 0:
            return default
        return json.loads(path.read_text(encoding="utf-8"))
    except Exception:
        return default


def write_json(path: Path, value) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(json.dumps(value, ensure_ascii=False, indent=2) + "\n", encoding="utf-8")


def fetch_json(url: str):
    request = urllib.request.Request(url, headers={"User-Agent": USER_AGENT})
    with urllib.request.urlopen(request, timeout=TIMEOUT) as response:
        return json.loads(response.read().decode("utf-8", "replace"))


def yahoo_chart(symbol: str):
    encoded = urllib.parse.quote(symbol, safe="")
    url = f"https://query1.finance.yahoo.com/v8/finance/chart/{encoded}?range=1mo&interval=1d&includePrePost=false"
    raw = fetch_json(url)
    result = (((raw or {}).get("chart") or {}).get("result") or [None])[0]
    if not result:
        raise RuntimeError(f"No chart result for {symbol}")
    timestamps = result.get("timestamp") or []
    quote = (((result.get("indicators") or {}).get("quote") or [None])[0] or {})
    closes = quote.get("close") or []
    points = []
    for ts, close in zip(timestamps, closes):
        if close is None:
            continue
        value = float(close)
        if not math.isfinite(value) or value <= 0:
            continue
        points.append([int(ts) * 1000, value])
    if len(points) < 3:
        raise RuntimeError(f"Insufficient points for {symbol}")
    return points[-31:]


def normalise(points):
    if not points:
        return []
    first = float(points[0][1])
    if not first:
        return []
    return [[point[0], round((float(point[1]) / first) * 10000) / 100] for point in points]


def daily_returns(points):
    values = [float(point[1]) for point in points]
    out = []
    for left, right in zip(values, values[1:]):
        if left:
            out.append((right / left) - 1)
    return out


def correlation(left, right):
    count = min(len(left), len(right))
    if count < 3:
        return None
    a = left[-count:]
    b = right[-count:]
    avg_a = sum(a) / count
    avg_b = sum(b) / count
    numerator = sum((x - avg_a) * (y - avg_b) for x, y in zip(a, b))
    var_a = sum((x - avg_a) ** 2 for x in a)
    var_b = sum((y - avg_b) ** 2 for y in b)
    denominator = math.sqrt(var_a * var_b)
    return round(numerator / denominator, 3) if denominator else None


def build_asset(key: str, previous_asset=None):
    meta = ASSETS[key]
    points = yahoo_chart(meta["symbol"])
    if meta.get("invert"):
        points = [[ts, 1 / value] for ts, value in points if value]
    values = [float(point[1]) for point in points]
    first = values[0]
    last = values[-1]
    last7 = values[-8] if len(values) > 8 else first
    return {
        "symbol": meta["symbol"],
        "label": meta["label"],
        "ticker": meta["ticker"],
        "unit": meta["unit"],
        "invert": bool(meta.get("invert")),
        "source": meta["source"],
        "current": round(last, 8),
        "change_7d_percent": round(((last / last7) - 1) * 100, 2) if last7 else None,
        "change_30d_percent": round(((last / first) - 1) * 100, 2) if first else None,
        "low_30d": round(min(values), 8),
        "high_30d": round(max(values), 8),
        "points": points,
        "indexed_points": normalise(points),
    }


def main() -> int:
    DATA.mkdir(parents=True, exist_ok=True)
    previous = read_json(MACRO_PATH, {})
    previous_assets = previous.get("assets") or {}
    assets = {}
    skipped = []
    successful = 0
    started = time.time()

    for key in ASSETS:
        try:
            assets[key] = build_asset(key, previous_assets.get(key))
            successful += 1
        except Exception as exc:
            # Private operational detail: keep the previous asset silently when available.
            if key in previous_assets:
                assets[key] = previous_assets[key]
            skipped.append({"asset": key, "reason": str(exc)[:160]})

    if not assets:
        raise RuntimeError("No macro market assets available")

    correlations = {}
    if "btc" in assets:
        btc_returns = daily_returns(assets["btc"].get("points") or [])
        for key in ("gold", "oil", "usd"):
            correlations[f"btc_{key}"] = correlation(btc_returns, daily_returns(assets.get(key, {}).get("points") or [])) if key in assets else None

    ts = now_iso()
    output = {
        "updated_at": ts,
        "title": "BTC, zlato, Brent nafta i USD/EUR - statistička usporedba",
        "period": "posljednjih 30 dostupnih dnevnih tržišnih vrijednosti",
        "disclaimer": "Indikativni javni tržišni podatci; prikaz nije usluga trgovanja niti investicijski savjet.",
        "source": "Public market data sources; unified portal refresh",
        "assets": assets,
        "correlations": correlations,
        "status": "ok" if successful >= 2 else "reference_checked",
        "successful_assets": successful,
        "configured_assets": len(ASSETS),
        "skipped_assets_private": skipped,
        "runtime_seconds": round(time.time() - started, 2),
    }
    write_json(MACRO_PATH, output)

    status = read_json(STATUS_PATH, {})
    status.update({
        "updated_at": ts,
        "cadence": status.get("cadence") or "scheduled every five minutes",
        "status": "ok",
        "macro_market": {
            "updated_at": ts,
            "status": output["status"],
            "assets": len(assets),
            "successful_assets": successful,
            "configured_assets": len(ASSETS),
            "public_message_policy": "temporarily unavailable macro sources are skipped silently when a valid package exists",
            "checked_at": ts,
            "last_attempt_at": ts,
            "last_successful_refresh_at": ts,
        },
    })
    write_json(STATUS_PATH, status)
    print(f"macro refresh: assets={len(assets)}, successful={successful}/{len(ASSETS)}")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: scripts/refresh-macro-market.mjs
import fs from 'node:fs';
import path from 'node:path';

const OUT = path.resolve('apps/portal/data/macro_market.json');

const ASSETS = [
  { key: 'btc', symbol: 'BTC-USD', label: 'Bitcoin', ticker: 'BTC', unit: 'USD / BTC', invert: false, needsRawPoints: true },
  { key: 'gold', symbol: 'GC=F', label: 'Zlato', ticker: 'XAU', unit: 'USD / oz', invert: false, needsRawPoints: false },
  { key: 'oil', symbol: 'BZ=F', label: 'Brent nafta', ticker: 'BRENT', unit: 'USD / barrel', invert: false, needsRawPoints: false },
  { key: 'usd', symbol: 'EURUSD=X', label: 'USD / EUR', ticker: 'USD', unit: 'EUR / USD', invert: true, needsRawPoints: false }
];

async function fetchYahooChart(symbol) {
  const url = `https://query1.finance.yahoo.com/v8/finance/chart/${encodeURIComponent(symbol)}?range=1mo&interval=1d`;
  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), 10000);
  try {
    const response = await fetch(url, {
      headers: { 'user-agent': 'Mozilla/5.0 (compatible; GNK-ASG-Macro-Refresh/1.0)' },
      signal: controller.signal
    });
    if (!response.ok) throw new Error(`HTTP ${response.status}`);
    const data = await response.json();
    const result = data?.chart?.result?.[0];
    if (!result) throw new Error('no chart result');
    const timestamps = result.timestamp || [];
    const closes = result.indicators?.quote?.[0]?.close || [];
    const points = [];
    for (let i = 0; i < timestamps.length; i++) {
      const close = closes[i];
      if (close == null || !Number.isFinite(Number(close))) continue;
      points.push([timestamps[i] * 1000, Number(close)]);
    }
    return points;
  } finally {
    clearTimeout(timer);
  }
}

function invertPoints(points) {
  return points.map(([t, v]) => [t, v ? 1 / v : v]);
}

function normalise(points) {
  if (!points.length || !Number(points[0][1])) return [];
  const first = Number(points[0][1]);
  return points.map(([t, v]) => [t, Math.round((Number(v) / first) * 10000) / 100]);
}

function dailyReturns(points) {
  return points.slice(1).map((row, index) => {
    const prev = Number(points[index][1]);
    return prev ? Number(row[1]) / prev - 1 : null;
  }).filter(v => v != null && Number.isFinite(v));
}

function correlation(a, b) {
  const count = Math.min(a.length, b.length);
  if (count < 3) return null;
  const left = a.slice(-count), right = b.slice(-count);
  const avgL = left.reduce((s, v) => s + v, 0) / count;
  const avgR = right.reduce((s, v) => s + v, 0) / count;
  const numerator = left.reduce((s, v, i) => s + (v - avgL) * (right[i] - avgR), 0);
  const varL = left.reduce((s, v) => s + (v - avgL) ** 2, 0);
  const varR = right.reduce((s, v) => s + (v - avgR) ** 2, 0);
  const denominator = Math.sqrt(varL * varR);
  return denominator ? Math.round((numerator / denominator) * 1000) / 1000 : null;
}

async function loadAsset(config) {
  try {
    let points = await fetchYahooChart(config.symbol);
    if (config.invert) points = invertPoints(points);
    if (points.length < 3) throw new Error('insufficient points');
    const values = points.map(p => p[1]);
    const last = values[values.length - 1];
    const first = values[0];
    const last7 = values.length > 8 ? values[values.length - 8] : first;
    const asset = {
      symbol: config.symbol,
      label: config.label,
      ticker: config.ticker,
      unit: config.unit,
      invert: config.invert,
      source: 'Yahoo Finance public chart' + (config.invert ? ', inverted EURUSD' : ''),
      current: last,
      change_7d_percent: Math.round(((last / last7) - 1) * 10000) / 100,
      change_30d_percent: Math.round(((last / first) - 1) * 10000) / 100,
      low_30d: Math.min(...values),
      high_30d: Math.max(...values),
      indexed_points: normalise(points)
    };
    if (config.needsRawPoints) asset.points = points;
    return { ok: true, key: config.key, asset, returns: dailyReturns(points) };
  } catch (error) {
    return { ok: false, key: config.key, error: String(error?.message || error) };
  }
}

async function main() {
  const started = Date.now();
  const results = await Promise.all(ASSETS.map(loadAsset));
  const assets = {};
  const returnsByKey = {};
  for (const result of results) {
    if (result.ok) {
      assets[result.key] = result.asset;
      returnsByKey[result.key] = result.returns;
    }
  }
  const successfulKeys = Object.keys(assets);
  const correlations = {};
  if (returnsByKey.btc) {
    for (const key of ['gold', 'oil', 'usd']) {
      correlations[`btc_${key}`] = returnsByKey[key] ? correlation(returnsByKey.btc, returnsByKey[key]) : null;
    }
  }
  const failed = results.filter(r => !r.ok);
  const dataset = {
    updated_at: new Date().toISOString(),
    title: 'BTC, zlato, Brent nafta i USD/EUR - statistička usporedba',
    period: 'posljednjih 30 dostupnih dnevnih tržišnih vrijednosti',
    disclaimer: 'Indikativni javni tržišni podatci; prikaz nije usluga trgovanja niti investicijski savjet.',
    source: 'Public market data sources; unified portal refresh',
    assets,
    correlations,
    status: successfulKeys.length === ASSETS.length ? 'ok' : (successfulKeys.length ? 'partial' : 'failed'),
    successful_assets: successfulKeys.length,
    configured_assets: ASSETS.length,
    skipped_assets_private: [],
    failed_assets: failed.map(f => ({ key: f.key, error: f.error })),
    runtime_seconds: Math.round(((Date.now() - started) / 1000) * 100) / 100
  };
  // Never overwrite good data with an empty/failed refresh; keep the previous file if nothing came back.
  if (successfulKeys.length === 0 && fs.existsSync(OUT)) {
    console.log(JSON.stringify({ ok: false, message: 'All asset fetches failed; previous macro_market.json left unchanged.', failed: dataset.failed_assets }, null, 2));
    process.exitCode = 1;
    return;
  }
  fs.writeFileSync(OUT, JSON.stringify(dataset, null, 2));
  console.log(JSON.stringify({ ok: true, status: dataset.status, successful_assets: dataset.successful_assets, failed_assets: dataset.failed_assets }, null, 2));
}

await main();
